sharedbuffers/mapped_struct.py

```diff
@@ -320,6 +320,342 @@ def unpack_from(cls, buf, offs, idmap = None, array = array.array, itemsizes = {
+                raise ValueError("Inconsistent data, unknown type code %r" % (dcode,))
+        else:
+            res = _struct.unpack(self.buf[obj_offs:obj_offs + itemsize])[0]
+
+        self.idmap[obj_offs] = res
+        return res
+
+    if cython.compiled:
+        @cython.locals(index = int)
+        def  __richcmp__(self, other, op):
+
+            if not islist(self) or not islist(other):
+                if op == Py_EQ:
+                    return False
+                elif op == Py_NE:
+                    return True
+                else:
+                    raise NotImplementedError
+
+            _cmp = proxied_list_cmp(self, other)
+            if op == Py_EQ:
+                return _cmp == 0
+            elif op == Py_NE:
+                return _cmp != 0
+            elif op == Py_LT:
+                return _cmp < 0
+            elif op == Py_LE:
+                return _cmp <= 0
+            elif op == Py_GT:
+                return _cmp > 0
+            else: # op == Py_GE:
+                return _cmp >= 0
+    else:
+        def __cmp__(self, other):
+            if not islist(other):
+                raise NotImplementedError
+            return proxied_list_cmp(self, other)
+
+        def _ne(self, other):
+            if not islist(other):
+                return True
+            return proxied_list_cmp(self, other) != 0
+
+        def _eq(self, other):
+            if not islist(other):
+                return False
+            return proxied_list_cmp(self, other) == 0
+
+    @cython.ccall
+    @cython.locals(offs = cython.ulonglong, dcode = cython.char, pbuf = 'const char *', objlen = cython.ulonglong)
+    def _cobjlen(self):
+
+        offs = self.offs
+        pbuf = cython.cast(cython.p_char, self.pybuf.buf)
+        dcode = pbuf[offs]
+
+        if dcode in ('B','b','H','h','I','i'):
+            objlen = cython.cast(cython.p_uint, pbuf + offs)[0] >> 8
+            if objlen == 0xFFFFFF:
+                objlen = cython.cast(cython.p_longlong, pbuf + offs + 4)[0]
+
+        elif dcode in ('q', 'd', 't'):
+            objlen = cython.cast(cython.p_longlong, pbuf + offs)[0] >> 8
+
+        else:
+            raise ValueError("Inconsistent data, unknown type code %r" % (dcode,))
+
+        return objlen
+
+    def _objlen(self):
+        offs = self.offs
+
+        buf = self.buf
+        dcode = buf[offs]
+
+        if dcode in ('B','b','H','h','I','i'):
+            objlen, = struct.unpack('<I', buf[offs+1:offs+4] + '\x00')
+            if objlen == 0xFFFFFF:
+                objlen = struct.unpack_from('<Q', buf, offs + 4)
+
+        elif dcode in ('q', 'd', 't'):
+            objlen, = struct.unpack('<Q', buf[offs+1:offs+8] + '\x00')
+        else:
+            raise ValueError("Inconsistent data, unknown type code %r" % (dcode,))
+
+        return objlen
+
+    def __len__(self):
+        if cython.compiled:
+            return self._cobjlen()
+        else:
+            return self._objlen()
```

Comment:
Again, don't have 2 versions of the same method, just have the if cython.compiled inside the method.